Updated contents and structure (#97)
matter-labs/zksync-web-v2-docs: docs/dev/fundamentals/zkSync.md

# Introduction to zkSync

## Perequisite

> If you are unfamiliar with rollups, you should cover the [rollups basics](./rollups.md) and read about ZK-rollups and Optimistic rollups, before learning about zkSync

Reviewer: zk-rollups or zk rollups?
Optimistic or optimistic?

Reviewer: period at the end

**zkSync** is a [ZK rollup](./rollups.md) trustless protocol that utilises zero-knowledge proofs to provide scalable low-cost transactions on Ethereum. 
zkSync ensures that all assets are stored in a single smart contract on the mainchain, while computation and storing data are performed off-chain. As all transactions are proven on the Ethereum mainchain, users enjoy the same security level as in Ethereum.

zkSync 2.0 is made to look and feel like Ethereum, but with lower fees. Just like on Ethereum, smart contracts are written in Solidity/Vyper and can be called using the same clients as the other EVM-compatible chains.

> zkSync doesn't require you to register a separate private key to use before usage, it supports existing Ethereum wallets work out of the box.

Reviewer: ```suggestion
zkSync 2.0 doesn't require you to register a separate private key to use before usage, it supports existing Ethereum wallets which work out of the box.

zkSync ensures that all assets are stored in a single smart contract on the mainchain, while computation and storing data are performed off-chain. As all transactions are proven on the Ethereum mainchain, users enjoy the same security level as in Ethereum.

zkSync 2.0 is made to look and feel like Ethereum, but with lower fees. Just like on Ethereum, smart contracts are written in Solidity/Vyper and can be called using the same clients as the other EVM-compatible chains.

zkSync 2.0 doesn't require you to register a separate private key to use before usage, it supports existing Ethereum wallets which work out of the box.

At this time zkSync is solely run and operated by the zkSync team's servers, and is therefore centralized; however, this will be be transitioned to a decentralized system in the near future.

## zkSync in comparison

zkSync [stands out remarkably](https://blog.matter-labs.io/evaluating-ethereum-l2-scaling-solutions-a-comparison-framework-b6b2f410f955) in security and usability among existing L2 scaling solutions. 

> Thanks to the combination of cutting-edge cryptography and on-chain data availability, zkRollup (the core network of zkSync) is the only L2 scaling solution that doesn't require any operational activity to keep the funds safe. 

Reviewer 2: > zkRollup (the core network of zkSync)

But we have already described above that zkSync is a zk rollup :)

If you were talking about zkRollup as the antonym for zkPorter from [here](https://blog.matter-labs.io/zkporter-a-breakthrough-in-l2-scaling-ed5e48842fbf), then I think it would require prior introduction to what it is, etc. It is better to talk only about zkSync as whole zkRollup for now

Additionally, it is only possible to deploy smart contracts with this type of transaction.

Knowing the details about the transaction format is not required to use zkSync's SDK, but if you are curious, you can read more about it [here](../../api/api.md#eip712).

## Confirmations and finality

On zkSync, each transaction is at one of the four stages:

- `Pending`: The transaction was received by the operator, but it has not been processed yet.
- `Processed`: The transaction is processed by the operator and is confirmed to be included in the next block.
- `Committed`: The transaction state diffs were published on Ethereum.
- `Finalized`: The SNARK validity proof for the transaction has been submitted and verified by the smart contract. After this step, the transaction is considered to be final.

The typical time for a transaction to go from `Processed` to `Finalized` is a couple of hours at the current stage.

Please note that for developer convenience, we usually treat the `Processed` and `Committed` states as a single stage called `Committed` since they have no difference from the UX/DexEx standpoints.

## zkSync features

> - ETH and ERC20 token transfers with instant confirmations and 10-minute finality on L1

Reviewer 2: ```suggestion
- ETH and ERC20 token transfers with instant confirmations and fast finality on L1
```

We aim for 10-minute finality, but this will likely not be the case from day 1, so it is better to avoid using these numbers

> - Transaction fees are extremely low (1/100th of the mainnet cost for ERC20 tokens and 1/30th for ETH transfers).

Reviewer 2: The numbers here are true for the [first](https://wallet.zksync.io/) version of zkSync. I would suggest to avoid using exact numbers before the release and we know the prices for sure

> - No registration is required to receive funds.

Reviewer 2: This was taken from the documentation of zkSync 1.x :) 

This point was emphasized in contract to some other zkRollups that required registration both to receive funds and send funds. On zkSync the users could receive funds without registration, but they still had to do a small registration transaction before they could actually do anything with these funds. So no registration to receive, but there is a registration to transact.

On zkSync 2.0 it is the same as on Ethereum or Optimistic rollups: no need to register at all. So probably it is better to remove this point altogether

- Payments to existing Ethereum addresses (including smart contracts) can be conveniently paid with the token being transferred.

> - Withdrawals to the mainnet take ~10 minutes.

Reviewer 2: I would suggest removing this point altogether for the same reasons as points above (until we've launched on mainnet I would prefer to not mention any concrete numbers)

- As with the Ethereum mainnet, no third parties are needed to keep funds safe.
- Users are always in control of their funds.
- zkSync is smart-contract friendly; it allows you to reuse existing Solidity code.
- zkSync is friendlier for exchanges.

> - zkSync has native support for NFTs.

Reviewer 2: This for the case for zkSync 1.x, but there is no *native* support for NFTs in zkSync 2.0. In zkSync 2.0, the users can write their own ERC721 contracts if they wish to, but there is no *native* support. 

So please remove this point

- There is no requirement for an operational activity to keep the funds safe.

### zkSync 2.0 highlights

- Mainnet-like security with zero reliance on 3rd parties.
- Permissionless EVM-compatible smart contracts.
- Standard Web3 API.
- Preserving key EVM features, such as smart contract composability.

> - Introducing new features, such as account abstraction and meta-transactions.

Reviewer 2: ```suggestion
- Introducing new features, such as account abstraction.
```

We don't have meta-transactions per se, the users could implement something like them if they wish to via account abstraction.

At this time, zkSync is solely run and operated by the zkSync team's servers and is therefore centralized. However, this will be transitioned to a decentralized system shortly.

## zkSync in comparison

zkSync [stands out remarkably](https://blog.matter-labs.io/evaluating-ethereum-l2-scaling-solutions-a-comparison-framework-b6b2f410f955) in security and usability among existing L2 scaling solutions. 

Both parts will be able to work with each other and be put together. This means that contracts and accounts on the zkRollup side will be able to work with accounts on the zkPorter side without any problems, and vice versa.

The general rollup workflow is as follows:

- Users can become owners in a rollup by depositing assets from L1 or receiving a transfer from other owners.
- Owners can transfer assets to each other.
- Owners can withdraw assets under their control to an L1 address.
Rollup operation requires the assistance of an operator, who rolls transactions together, computes a zero-knowledge proof of the correct state transition, and affects the state transition by interacting with the rollup contract.

> To understand the design, we need to look into how zkSync rollup transactions.

Reviewer 2: "into how zkSync rollup transactions."? you probably missed a verb here

zkSync operations are divided into rollup transactions (initiated inside Rollup by a Rollup account) and priority operations (initiated on the mainchain by an Ethereum account).

> The zkSync rollup operation lifecycles are as follows;

Reviewer 2: What is the difference between "Rollup" and "rollup" here?

- The proof for the block is submitted to the zkSync smart contract as block verification. If the verification succeeds, the new state is considered finalized.

Furthermore, on zkSync, each L2 block will progress through the following four stages until it is finalized.

- `Pending`: The transaction was received by the operator, but it has not been processed yet.
- `Processed`: The transaction is processed by the operator and is confirmed to be included in the next block.
committed indicates that the transaction data of this block has been posted on Ethereum. It does not prove that it has been executed in a valid way, but it ensures the availability of the block data.
- `Finalized`: This indicates that the SNARK validity proof for the transaction has been submitted and verified by the smart contract. After this step, the transaction is considered to be final.

The typical time for a transaction to go from `Processed` to `Finalized` is a couple of hours at the current stage.

Please note that for developer convenience, we usually treat the `Processed` and `Committed` states as a single stage called `Committed` since they have no difference from the UX/DexEx standpoints.

> ### The State of the Rollup

Reviewer 2: What is meant here by "Rollup" with capitalized "R"?

The current version of zkSync 2.0 solves the needs of most applications on Ethereum, and with more features planned for release soon, zkSync 2.0 will provide developers with a design space to experiment with applications not possible on Ethereum today. With this release, we are supporting the following features:

- Native support of ECDSA signatures: Unlike the first version of zkSync and most of the ZK Rollups, no special operation is required to register the user’s private key. Any account can be managed in L2 with the same private key that is used for L1.
- Solidity 0.8.x support: Deploy your existing codebase with little to no changes required.
Web3 API. With small exceptions, our API is fully compatible with Ethereum. This allows seamless integration with existing indexers, explorers, etc.

> - Support for Ethereum cryptographic primitives: zkSync natively supports `keccak256`, `sha256`, and `ecrecover` via precompilers.

Reviewer 2: ```suggestion
- Support for Ethereum cryptographic primitives: zkSync natively supports `keccak256`, `sha256`, and `ecrecover` via precompiles.
```

- Hardhat plugin: Allows easy testing and development of smart contracts on zkSync.
- L1 → L2 smart contract messaging: Allows developers to pass data from Ethereum L1 to smart contracts on zkSync, providing the required information to run various smart contracts.
Some features are not included in our current testnet that we’re looking to ship in future upgrades. In estimated chronological order, this includes:

> - Account abstraction: Imagine being able to implement custom logic for signature checking for your account. Or maybe social recovery? Currently, on most of the EVM chains, users need to deploy smart contract wallets for such purposes. All of this would be easily supported with account abstraction.

Reviewer 2: Account abstraction is already included. Even though some breaking changes are possible, it is included already

> - Support for older versions of Solidity. We are working hard to support different versions of Solidity so that new projects can be added without any problems.

Reviewer 2: In the rother part of the documentation we already stated that we support Solidity 0.4.x

zkSync [stands out remarkably](https://blog.matter-labs.io/evaluating-ethereum-l2-scaling-solutions-a-comparison-framework-b6b2f410f955) in security and usability among existing L2 scaling solutions. 

> Both parts will be able to work with each other and be put together. This means that contracts and accounts on the zkrollup side will be able to work with accounts on the zkPorter side without any problems, and vice versa.

Reviewer 2: the same here

The general rollup workflow is as follows:

- Users can become owners in a rollup by depositing assets from L1 or receiving a transfer from other owners.
- Owners can transfer assets to each other.
- Owners can withdraw assets under their control to an L1 address.

> rollup operation requires the assistance of an operator, who rolls transactions together, computes a zero-knowledge proof of the correct state transition, and affects the state transition by interacting with the rollup contract.

Reviewer 2: This is the first word in a sentence, so it should be capitalized

> To understand the design, we need to look into how zkSync rollup transactions works.

Reviewer 2: "work".

Since we are talking plural, we should use "work" and not works

Please note that for developer convenience, we usually treat the `Processed` and `Committed` states as a single stage called `Committed` since they have no difference from the UX/DexEx standpoints.

> ### The State of the rollup

Reviewer 2: Let's use "testnet" or "zkSync" here

- Hardhat plugin: Allows easy testing and development of smart contracts on zkSync.
- L1 → L2 smart contract messaging: Allows developers to pass data from Ethereum L1 to smart contracts on zkSync, providing the required information to run various smart contracts.
Some features are not included in our current testnet that we’re looking to ship in future upgrades. In estimated chronological order, this includes:

- zkPorter: One of the largest and most important features, zkPorter will allow users to choose between a zkrollup account featuring the highest security and a 20x fee reduction compared to Ethereum, or a zkPorter account featuring stable transaction fees of just a few cents in a different security model (much higher than that of a sidechain). Both zkPorter accounts, and zkrollup accounts, will be able to interact seamlessly together under the hood.

## zkSync in comparison

zkSync [stands out remarkably](https://blog.matter-labs.io/evaluating-ethereum-l2-scaling-solutions-a-comparison-framework-b6b2f410f955) in security and usability among existing L2 scaling solutions.

> Thanks to the combination of cutting-edge cryptography and on-chain data availability, zkrollup (the core network of zkSync) are the only L2 scaling solution that doesn't require any operational activity to keep the funds safe.

Reviewer 2: zkRollup

> For example, users can go offline and still be able to withdraw their assets safely when they come back, even if the zkrollup validators are no longer around.

Reviewer 2: zkRollup
